src/main/java/com/google/sps/servlets/queue/EnterQueue.java

```diff
@@ -85,82 +85,87 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
-            ZoneId defaultZoneId = ZoneId.systemDefault();
-            Date currDate = Date.from(localDate.atStartOfDay(defaultZoneId).toInstant());
-
-            // Get time
-            LocalDateTime localTime = LocalDateTime.now(clock);
-            Date currTime = Date.from(localTime.atZone(defaultZoneId).toInstant());
-
-            // Query visit entity for particular day
-            Filter classVisitFilter =
-                new FilterPredicate("classKey", FilterOperator.EQUAL, classKey);
-            Filter dateVisitFilter = new FilterPredicate("date", FilterOperator.EQUAL, currDate);
-            CompositeFilter visitFilter =
-                CompositeFilterOperator.and(dateVisitFilter, classVisitFilter);
-            PreparedQuery query = datastore.prepare(new Query("Visit").setFilter(visitFilter));
-
-            // Get visit entity for particular day
-            Entity visitEntity;
-            if (query.countEntities() == 0) {
-              visitEntity = new Entity("Visit");
-              visitEntity.setProperty("classKey", classKey);
-              visitEntity.setProperty("date", currDate);
-              visitEntity.setProperty("numVisits", (long) 0);
-            } else {
-              visitEntity = query.asSingleEntity();
-            }
-
-            long numVisits = (long) visitEntity.getProperty("numVisits");
-            List<EmbeddedEntity> updatedQueue =
-                (List<EmbeddedEntity>) classEntity.getProperty("studentQueue");
-
-            // Update studentQueue and numVisit properties if student not already in queue
-            if (!updatedQueue.stream()
-                .filter(elem -> (((String) elem.getProperty("uID")).equals(userID)))
-                .findFirst()
-                .isPresent()) {
-
-              // create new student entity to add to queue
-              EmbeddedEntity studentInfo = new EmbeddedEntity();
-              studentInfo.setProperty("timeEntered", currTime);
-
-              Workspace w = workspaceFactory.create(classCode);
-              w.setStudentUID(userID);
-              studentInfo.setProperty("workspaceID", w.getWorkspaceID());
-
-              studentInfo.setProperty("uID", userID);
 
-              updatedQueue.add(studentInfo);
-              numVisits++;
-            }
-
-            visitEntity.setProperty("numVisits", numVisits);
-            datastore.put(txn, visitEntity);
-
-            classEntity.setProperty("studentQueue", updatedQueue);
-            datastore.put(txn, classEntity);
-
-            txn.commit();
-            break;
-          } catch (ConcurrentModificationException e) {
-            if (retries == 0) {
-              throw e;
-            }
-            // Allow retry to occur
-            --retries;
-          } finally {
-            if (txn.isActive()) {
-              txn.rollback();
+        EmbeddedEntity beingHelped =
+            (EmbeddedEntity) datastore.get(classKey).getProperty("beingHelped");
```

Comment:
This should be in the transaction.

Reply:
it is in transaction!

Reply:
It looks like the transaction is created below this on line `95` right?

Reply:
ah okay, I see now!